### PR TITLE
bpo-34060: Report system load when running test suite for Windows

### PR DESCRIPTION
**Pull request by ammaraskar**
See the bpo ticket for more technical details on the implementation and otherwise.

https://bugs.python.org/issue34060

### CODE REVIEW

**Comment by jkloth** [comment]
A quick note on the LOADAVG_FACTOR_1F and SAMPLING_INTERVAL; that given dampening factor is for a sample rate of 5 seconds.  SAMPLING_INTERVAL should really be changed to 5 or re-run the equation with the different interval.

**Comment by ammaraskar** [comment]
Thanks again Jeremy, I changed the interval to 5 to be more in line with the Linux scheduler.

File: Modules/_winapi.c [hunk 1]
```diff
@@ -1699,6 +1700,92 @@ _winapi_GetFileType_impl(PyObject *module, HANDLE handle)
...
+
+static PyObject *
+_winapi_InitializeLoadCounter_impl(PyObject *module)
+/*[clinic end generated code: output=943e4187b2a20b39 input=3caa0958b4abcea0]*/
+{
+    PDH_STATUS s;
+    HQUERY hQuery;
+    HCOUNTER hCounter;
+
+    if ((s = PdhOpenQueryW(NULL, 0, &hQuery)) != ERROR_SUCCESS)
```
Comment (vstinner) [inline Modules/_winapi.c:1755]: Please move the assignment on a separated line. Same comments for other assignments in if().

File: Modules/_winapi.c [hunk 2]
```diff
@@ -1699,6 +1700,92 @@ _winapi_GetFileType_impl(PyObject *module, HANDLE handle)
...
+        goto WMIerror;
+    }
+
+    WCHAR *szCounterPath = L"\\System\\Processor Queue Length";
+    if ((s = PdhAddEnglishCounterW(hQuery, szCounterPath, 0, &hCounter)) != ERROR_SUCCESS)
+    {
+        goto WMIerror;
+    }
+
+    HANDLE event = CreateEventW(NULL, FALSE, FALSE, L"LoadUpdateEvent");
```
Comment (vstinner) [inline Modules/_winapi.c:1766]: Hum. It seems like you allocate a resource, but never release it.

I would suggest to create an object which keeps track of these resources and release them later. getloadavg() would be a method of that object.

Comment (ammaraskar) [reply Modules/_winapi.c:1766]: Would it be acceptable to do it in the module's `m_free` function? I was initially thinking of doing something similar but thought that would create a lot of extra code and complexity for an internal API with one use case.

**Comment by ammaraskar** [comment]
Also, I don't think this actually needs a NEWS entry right? It relates to cpython build process and doesn't actually affect the language. Could you mark it with the `skip news` label if so?

File: Modules/_winapi.c (hunk 2 above)
Comment (vstinner) [reply Modules/_winapi.c:1766]: I dislike using m_free() for that.

**Comment by zooba** [comment]
What extra runtime dependencies does this add to `python38.dll` (use `dumpbin /imports python38.dll` and grab the DLL filenames)?

Also, `winapi` methods are public, if undocumented (since they are supposed to reflect the Windows API calls directly, so that documentation should suffice). If you really want an internal test suite helper it should go in its own extension module with a `_` prefix.

(Haven't looked too closely at the implementation yet, but when I do, it will be with the view that it is a public API change. Unless you rework it into its own module first.)

**Comment by zooba** [comment]
(Edited above post). Let me take that back - was thinking of `winreg` module, not the `_winapi` module. But it still should be treated as a publicly usable API with all the robustness guarantees unless it's in its own module, since it will be used.

**Comment by ammaraskar** [comment]
Aah yeah I think you have a good point, `_winapi` despite being undocumented is being consumed by end users https://github.com/search?p=1&q=import+_winapi&type=Code

I think making it its own module would be a bit overkilll so I'll clean up the API anticipating some public usage.

**Comment by ammaraskar** [comment]
Closed in favor of https://github.com/python/cpython/pull/8357

Thank you @csabella 😄

**Pull request closed without merging**